8285973: x86_64: Improve fp comparison and cmove for eq/ne

merykitty opened the pull request:
Hi,

This patch optimises the matching rules for floating-point comparison with respects to eq/ne on x86-64

1, When the inputs of a comparison is the same (i.e `isNaN` patterns), `ZF` is always set, so we don't need `cmpOpUCF2` for the eq/ne cases, which improves the sequence of `If (CmpF x x) (Bool ne)` from

    ucomiss xmm0, xmm0
    jp      label
    jne     label

into

    ucomiss xmm0, xmm0
    jp      label

2, The move rules for `cmpOpUCF2` is missing, which makes patterns such as `x == y ? 1 : 0` to fall back to `cmpOpU`, which have a really high cost of fixing the flags, such as

        xorl    ecx, ecx
        ucomiss xmm0, xmm1
        jnp     done
        pushf
        andq    [rsp], 0xffffff2b
        popf
    done:
        movl    eax, 1
        cmovel  eax, ecx

The patch changes this sequence into

    xorl    ecx, ecx
    ucomiss xmm0, xmm1
    movl    eax, 1
    cmovpl  eax, ecx
    cmovnel eax, ecx

3, The patch also changes the pattern of `isInfinite` to be more optimised by using `Math.abs` to reduce 1 comparison and compares the result with `MAX_VALUE` since `>` is more optimised than `==` for floating-point types.

The benchmark results are as follow:

                                                     Before              After
    Benchmark                      Mode  Cnt     Score     Error    Score     Error   Unit   Ratio
    FPComparison.equalDouble       avgt    5  2876.242 ±  58.875  594.636 ±   8.922  ns/op    4.84
    FPComparison.equalFloat        avgt    5  3062.430 ±  31.371  663.849 ±   3.656  ns/op    4.61
    FPComparison.isFiniteDouble    avgt    5   475.749 ±  19.027  518.309 ± 107.352  ns/op    0.92
    FPComparison.isFiniteFloat     avgt    5   506.525 ±  14.417  515.576 ±  14.669  ns/op    0.98
    FPComparison.isInfiniteDouble  avgt    5  1232.800 ±  31.677  621.185 ±  11.935  ns/op    1.98
    FPComparison.isInfiniteFloat   avgt    5  1234.708 ±  70.239  623.566 ±  15.206  ns/op    1.98
    FPComparison.isNanDouble       avgt    5  2255.847 ±   7.238  400.124 ±   0.762  ns/op    5.64
    FPComparison.isNanFloat        avgt    5  2567.044 ±  36.078  546.486 ±   1.509  ns/op    4.70

Thank you very much.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8285973](https://bugs.openjdk.java.net/browse/JDK-8285973): x86_64: Improve fp comparison and cmove for eq/ne


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8525/head:pull/8525` \
`$ git checkout pull/8525`

Update a local copy of the PR: \
`$ git checkout pull/8525` \
`$ git pull https://git.openjdk.java.net/jdk pull/8525/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8525`

View PR using the GUI difftool: \
`$ git pr show -t 8525`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8525.diff">https://git.openjdk.java.net/jdk/pull/8525.diff</a>

</details>

vnkozlov commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -6970,6 +6972,34 @@ instruct cmovI_regUCF(cmpOpUCF cop, rFlagsRegUCF cr, rRegI dst, rRegI src) %{
+            "cmovnel $dst, $src" %}
+  ins_encode %{
+    __ cmovl(Assembler::parity, $dst$$Register, $src$$Register);
+    __ cmovl(Assembler::notEqual, $dst$$Register, $src$$Register);
```
Should this be `equal`?

vnkozlov replied:
I see that you swapped `src, dst` in `match()` but `format` is sill incorrect and the code is confusing.

merykitty replied:
This is a flipping the sense of the test by flipping the input of the `CMove`, so this is essentially the same as in the above rule. Thanks.

sviswa7 replied:
I agree with @vnkozlov that this needs explanation. Could you please add comments here with IR and example code generated for both the eq case and ne case? You have some explanation in the PR description but not in the code. The description needs to be in the code as well for maintenance.

vnkozlov replied:
Right, I missed that. Then you can use `expand %{` to avoid duplication (keep format).

vnkozlov reviewed (changes requested):
Thank you for including tests. But you need additional other float compare (not just `ea` `ne`) tests since you removed some of `Cmp` instructions.

You need approval from core libs for Java methods changes (it affects all platforms). And we will get intrinsics for them soon (I think): #8459.

Please add comments to `cmpOpU*` operands explaining changes in them.

You did not explain removal of some float compare instructions.

PaulSandoz reviewed (commented):
The changes to `Float` and `Double` look good. I don't think we need additional tests, see test/jdk/java/lang/Math/IeeeRecommendedTests.java.

At first i thought we no longer need PR #8459 but it seems they are complimentary, albeit with more modest performance gains for the intrinsics.

vnkozlov commented:
> The changes to `Float` and `Double` look good. I don't think we need additional tests, see test/jdk/java/lang/Math/IeeeRecommendedTests.java.

Thank you, Paul for pointing the test. It means we need to run tier4 (which runs these tests with -Xcomp) to make sure methods are compiled by C2.

merykitty commented:
I have reverted the changes to `java.lang.Float` and `java.lang.Double` to not interfere with the intrinsic PR. More tests are added to cover all cases regarding floating-point comparison of compiled code.

The rules for fp comparison that output the result to `rFlagRegsU` are expensive and should be avoided. As a result, I removed the shortcut rules with memory or constant operands to reduce the number of match rules. Only the basic rules are kept.

Thanks.

vamsi-parasa commented:
Could you pls show the performance delta with the baseline and after the patch? Otherwise, people reviewing this PR have to manually compute how much improvement is obtained. For example, `FPComparison.isNanFloat` is showing `4.7x` improvement. Kindly fill the delta column for rest of the data points. 

Instead of two separate tables, the suggested table format for each row would be:
`<benchmark> , <baseline throughput> , <this patch throughput>, <performance delta in % or x>`

sviswa7 commented:
@merykitty Very nice work! The patch looks good to me.

sviswa7 commented:
@merykitty Very nice work! The patch looks good to me.

merykitty commented:
@vnkozlov  I have added comments to describe the changes in `cmpOpUCF` and the reasons behind the `cmov_regUCF2_eq` match rules. Using `expand` broke the build with `Syntax Error: :For expand in cmovI_regUCF2_eq to work, parameter declaration order in cmovI_regUCF2_ne must follow matchrule`.

@sviswa7  (x != y) ? a : b can be calculated using pseudocode as follow:

    res = (!ZF || PF) ? a : b
        = !ZF ? a : (PF ? a : b)

which can be calculated using

    cmovp  rb, ra // rb1 = PF ? ra : rb
    cmovne rb, ra // rb2 = !ZF ? ra : rb1 = !ZF ? ra : (PF ? ra : rb)

Furthermore, since `(x == y) == !(x != y)`, we have `((x == y) ? a : b) == ((x != y) ? b : a)`, which explains the implementation of `cmov_regUCF2_eq`.

@vamsi-parasa Thanks a lot for your suggestion, I have modified the PR description as you say.

vnkozlov reviewed (approved):
Thank you for trying my suggestion and new comments. Approved.

vnkozlov commented:
I started our testing. Please, wait results.

vnkozlov commented:
> I started our testing. Please, wait results.

Testing passed clean.

merykitty commented:
Thank you very much for your reviews and testing.
/integrate

sviswa7 commented:
/integrate

sviswa7 commented:
/sponsor
